client/allocrunner/taskrunner/artifact_hook.go

```diff
@@ -42,7 +42,10 @@ func (h *artifactHook) Prestart(ctx context.Context, req *interfaces.TaskPrestar
 	for _, artifact := range req.Task.Artifacts {
 		//XXX add ctx to GetArtifact to allow cancelling long downloads
 		if err := getter.GetArtifact(req.TaskEnv, artifact, req.TaskDir.Dir); err != nil {
-			wrapped := fmt.Errorf("failed to download artifact %q: %v", artifact.GetterSource, err)
+			wrapped := structs.NewRecoverableError(
+				fmt.Errorf("failed to download artifact %q: %v", artifact.GetterSource, err),
```

Comment:
❓ Could we potentially retry these here, rather than restarting the whole task? It feels like it would be much cheaper and prevent us from redownloading potentially expensive artifacts.

Reply:
We want to use the task restart machinery so the `restart` policy is applied.

Most prestart hooks will noop on restart if they've already run successfully. However, if multiple artifacts are being downloaded and the last one fails, this will download all of the artifacts all over again which is wasteful.

We could mark each artifact done individually to track this. I'll quick see if the implementation is something I feel comfortable sneaking in this late in 0.9 :grin:

Reply:
Pushed this fix. The HookData system is a bit awkward, but I think it worked out ok.